### PR TITLE
Tracing run perf report support

### PR DESCRIPTION
**Pull request by mo-tenstorrent**
### Ticket
#9956 

### Problem description
In trace runs, there are more device ops than host issued ops. Trace runs cannot generate host + device op report 

### What's changed
Track issuing of traces and use host runtime op id to attach host op id to tracing device ops as well.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11599915728
- [x] Device perf https://github.com/tenstorrent/tt-metal/actions/runs/11616448537 Broken with a known issue
- [x] T3K profiler https://github.com/tenstorrent/tt-metal/actions/runs/11597622995/job/32291710908

Sample report for a 2CQ trace resnet performant run on T3K.
Note the `TRACE ID` and `TRACE RUNTIME ID` columns after the IO columns.

TRACE ID is what tt_metal has as assigned for the trace id.
TRACE RUNTIME ID is generated by profiler as the count on how many times that trace ID ran. This is extracted from counting the op series.

[ops_perf_results_2024_11_01_14_25_55.csv](https://github.com/user-attachments/files/17600556/ops_perf_results_2024_11_01_14_25_55.csv)

### CODE REVIEW

File: tt_metal/tools/profiler/tt_metal_tracy.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Comment (tt-rkim) [inline tt_metal/tools/profiler/tt_metal_tracy.hpp:1]: Do we need a `#pragma once` ?

**Review by tt-rkim — APPROVED** [review]

**Comment by ayerofieiev-tt** [comment]
Can you share a sample trace after this change?

File: tt_metal/tools/profiler/tt_metal_tracy.hpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
```
Comment (tt-aho) [inline tt_metal/tools/profiler/tt_metal_tracy.hpp:1]:
```suggestion
// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```

File: tests/scripts/run_profiler_regressions.sh [hunk 3]
```diff
@@ -4,45 +4,69 @@ source scripts/tools_setup_common.sh
...
+        env WH_ARCH_YAML=wormhole_b0_80_arch_eth_dispatch.yaml ./tt_metal/tools/profiler/profile_this.py -c "pytest models/demos/t3000/resnet50/tests/test_resnet50_performant.py::test_run_resnet50_trace_2cqs_inference[wormhole_b0-True-True-16-act_dtype0-weight_dtype0-math_fidelity0-device_params0]" | tee $PROFILER_ARTIFACTS_DIR/test_out.log
 
         if cat $PROFILER_ARTIFACTS_DIR/test_out.log | grep "SKIPPED"
         then
             echo "No verification as test was skipped"
         else
             echo "Verifying test results"
             runDate=$(ls $PROFILER_OUTPUT_DIR/)
-            LINE_COUNT=1000 # Smoke test to see at least 1000 ops are reported
+            LINE_COUNT=4100 # Smoke test to see at least 1000 ops are reported
```
Comment (tt-aho) [inline tests/scripts/run_profiler_regressions.sh:43]: Comment should be 4100 now?

**Review by tt-aho — APPROVED** [review]

**Review by pgkeller — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

File: tt_metal/impl/device/device.cpp [hunk 4]
```diff
@@ -3299,6 +3300,8 @@ bool Device::using_slow_dispatch() const {
 }
 
 void Device::begin_trace(const uint8_t cq_id, const uint32_t tid) {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/device/device.cpp:3302]: Do you think this must be renamed to "start_replay_capture" and "end_replay_capture"?

**Pull request merged**